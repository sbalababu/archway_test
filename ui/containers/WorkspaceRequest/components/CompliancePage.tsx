import * as React from 'react';
import { Col, Row } from 'antd';
import ComplianceCard from './ComplianceCard';
import { RequestInput } from '../../../models/RequestInput';

const complianceData = {
  pci: {
    icon: 'bank',
    label: 'Payment Card Industry [Data Security Standard]',
    values: [
      {
        key: 'card',
        label: 'Full or partial credit card numbers?',
      },
      {
        key: 'bank',
        label: 'Full or partial bank account numbers?',
      },
      {
        key: 'other',
        label: 'Any other combination of data that can be used to make purchases?',
      },
    ],
  },
  pii: {
    icon: 'idcard',
    label: 'Personally Identifiable Information',
    values: [
      {
        key: 'name',
        label: 'Full name',
      },
      {
        key: 'address',
        label: 'Home address',
      },
      {
        key: 'email',
        label: 'Email address',
      },
      {
        key: 'ssn',
        label: 'Social security number',
      },
      {
        key: 'passport',
        label: 'Passport number',
      },
      {
        key: 'driver_license',
        label: `Driver's license number`,
      },
      {
        key: 'credit_card',
        label: 'Credit card number',
      },
      {
        key: 'dob',
        label: 'Date of birth',
      },
      {
        key: 'phone',
        label: 'Telephone number',
      },
      {
        key: 'credentials',
        label: 'Software credentials',
      },
    ],
  },
  phi: {
    icon: 'medicine-box',
    label: 'Protected Health Information',
    values: [
      {
        key: 'status',
        label: 'Health status',
      },
      {
        key: 'provision',
        label: 'Provision of health care',
      },
      {
        key: 'payment',
        label: 'Payment for health care',
      },
    ],
  },
};

interface Props {
  request: RequestInput;
  setRequest: (request: RequestInput | boolean) => void;
}

interface State {
  pci: any;
  pii: any;
  phi: any;
}

class CompliancePage extends React.Component<Props, State> {
  public state: State = {
    pci: {},
    pii: {},
    phi: {},
  };

  public componentDidMount() {
    window.addEventListener('keyup', this.onKeyUp);
  }

  public componentWillUnmount() {
    window.removeEventListener('keyup', this.onKeyUp);
  }

  public getBooleanFromData(data: any) {
    return Object.keys(data).filter(key => !!data[key]).length > 0;
  }

  public onKeyUp: any = (e: React.KeyboardEvent) => {
    if (e.key === 'a') {
      const { request, setRequest } = this.props;

      ['pci', 'pii', 'phi'].map(type => {
        complianceData[type].values.map((value: any) => {
          const { key } = value;
          const { [type as ('pci' | 'pii' | 'phi')]: data } = this.state;
          const newData = {
            ...data,
            [key]: true,
          };
          switch (type) {
            case 'pci':
              this.setState({ pci: newData });
              break;
            case 'pii':
              this.setState({ pii: newData });
              break;
            case 'phi':
              this.setState({ phi: newData });
              break;
            default:
          }
        });
      });
      setRequest({
        ...request,
        compliance: {
          pci_data: true,
          pii_data: true,
          phi_data: true,
        },
      });
    }
  };

  public onChange(type: 'pci' | 'pii' | 'phi', name: string, value: boolean) {
    const { request, setRequest } = this.props;
    const { [type]: data } = this.state;
    const newData = {
      ...data,
      [name]: value,
    };
    switch (type) {
      case 'pci':
        this.setState({ pci: newData });
        break;
      case 'pii':
        this.setState({ pii: newData });
        break;
      case 'phi':
        this.setState({ phi: newData });
        break;
      default:
    }

    if (Object.keys(newData).length === complianceData[type].values.length) {
      let typeKey: 'pci_data' | 'pii_data' | 'phi_data' = 'pci_data';
      if (type === 'pii') {
        typeKey = 'pii_data';
      } else if (type === 'phi') {
        typeKey = 'phi_data';
      }
      setRequest({
        ...request,
        compliance: {
          ...request.compliance,
          [typeKey]: this.getBooleanFromData(newData),
        },
      });
    }
  }

  public render() {
    const { pci, pii, phi } = this.state;

    return (
      <div>
        <h3>Help us be proactive in addressing risk-related data needs</h3>
        <Row type="flex" gutter={25} style={{ marginTop: 25, marginBottom: 25 }}>
          <Col span={8}>
            <ComplianceCard
              type="pci"
              data={complianceData.pci}
              values={pci}
              onChange={(name: string, value: boolean) => this.onChange('pci', name, value)}
            />
          </Col>
          <Col span={8}>
            <ComplianceCard
              type="pii"
              data={complianceData.pii}
              values={pii}
              onChange={(name: string, value: boolean) => this.onChange('pii', name, value)}
            />
          </Col>
          <Col span={8}>
            <ComplianceCard
              type="phi"
              data={complianceData.phi}
              values={phi}
              onChange={(name: string, value: boolean) => this.onChange('phi', name, value)}
            />
          </Col>
        </Row>
      </div>
    );
  }
}

export default CompliancePage;
