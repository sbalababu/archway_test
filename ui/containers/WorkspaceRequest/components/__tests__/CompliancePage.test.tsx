import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import CompliancePage from '../CompliancePage';

describe('CompliancePage', () => {
  it('renders correctly', () => {
    const props = {
      request: {
        name: 'request',
        summary: 'summary',
        description: 'desc',
        compliance: {
          phi_data: true,
          pci_data: true,
          pii_data: true,
        },
      },
      setRequest: () => null,
    };
    const wrapper = shallow(<CompliancePage {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
