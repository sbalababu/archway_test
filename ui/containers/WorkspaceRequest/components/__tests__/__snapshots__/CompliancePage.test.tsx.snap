// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompliancePage renders correctly 1`] = `
<div>
  <h3>
    Help us be proactive in addressing risk-related data needs
  </h3>
  <Row
    gutter={25}
    style={
      Object {
        "marginBottom": 25,
        "marginTop": 25,
      }
    }
    type="flex"
  >
    <Col
      span={8}
    >
      <ComplianceCard
        data={
          Object {
            "icon": "bank",
            "label": "Payment Card Industry [Data Security Standard]",
            "values": Array [
              Object {
                "key": "card",
                "label": "Full or partial credit card numbers?",
              },
              Object {
                "key": "bank",
                "label": "Full or partial bank account numbers?",
              },
              Object {
                "key": "other",
                "label": "Any other combination of data that can be used to make purchases?",
              },
            ],
          }
        }
        onChange={[Function]}
        type="pci"
        values={Object {}}
      />
    </Col>
    <Col
      span={8}
    >
      <ComplianceCard
        data={
          Object {
            "icon": "idcard",
            "label": "Personally Identifiable Information",
            "values": Array [
              Object {
                "key": "name",
                "label": "Full name",
              },
              Object {
                "key": "address",
                "label": "Home address",
              },
              Object {
                "key": "email",
                "label": "Email address",
              },
              Object {
                "key": "ssn",
                "label": "Social security number",
              },
              Object {
                "key": "passport",
                "label": "Passport number",
              },
              Object {
                "key": "driver_license",
                "label": "Driver's license number",
              },
              Object {
                "key": "credit_card",
                "label": "Credit card number",
              },
              Object {
                "key": "dob",
                "label": "Date of birth",
              },
              Object {
                "key": "phone",
                "label": "Telephone number",
              },
              Object {
                "key": "credentials",
                "label": "Software credentials",
              },
            ],
          }
        }
        onChange={[Function]}
        type="pii"
        values={Object {}}
      />
    </Col>
    <Col
      span={8}
    >
      <ComplianceCard
        data={
          Object {
            "icon": "medicine-box",
            "label": "Protected Health Information",
            "values": Array [
              Object {
                "key": "status",
                "label": "Health status",
              },
              Object {
                "key": "provision",
                "label": "Provision of health care",
              },
              Object {
                "key": "payment",
                "label": "Payment for health care",
              },
            ],
          }
        }
        onChange={[Function]}
        type="phi"
        values={Object {}}
      />
    </Col>
  </Row>
</div>
`;
